src/compiler/transformers/utilities.ts

```diff
@@ -178,4 +178,17 @@ namespace ts {
         }
         return values;
     }
+
+    /**
+     * Used in the module transformer to check if an expression is reasonably without sideeffect,
+     *  and thus better to copy into multiple places rather than to cache in a temporary variable
+     *  - this is mostly subjective beyond the requirement that the expression not be sideeffecting
+     */
+    export function isSimpleCopiableExpression(expression: Expression) {
+        return expression.kind === SyntaxKind.StringLiteral ||
+            expression.kind === SyntaxKind.NumericLiteral ||
+            expression.kind === SyntaxKind.NoSubstitutionTemplateLiteral ||
+            isKeyword(expression.kind) ||
+            isEntityNameExpression(expression);

```

Comment:
Any property access can have side effects. I would only recommend reusing something that is a `PrimaryExpression` (e.g. not property-access/entity name).

Reply:
Teeeeeechnically even accessing an `Identifier` can have sideeffects within the context of a `with` statement. I get where you're coming from and will dial it back to just `Identifier`, though. The relevant bits of primary expression are all here, and there is not a `isPrimaryExpression` (and if there were, it'd include parenthesized expressions and object/array literals), so this is about as good as it gets.